### PR TITLE
Lock the retro when finish or locked by PO + timer for retro with socket IO

### PR DESCRIPTION
**Pull request by Kevw94**

### CODE REVIEW

File: back-end/src/common/gateways/retrospective.global.gateway.ts [hunk 1]
```diff
@@ -101,4 +101,82 @@ export class RetrospectiveGateway
...
+
+	@SubscribeMessage('end-currentRetro')
+	async endCurrentRetro(client: AuthSocket) {
+		const endedDate = new Date();
+		client.to(client.roomId).emit('end-currentRetro', endedDate);
+
+		const query = { slug: client.roomId };
+		const update = {$set: { endedAt: endedDate, isRetroEnded: true}}
+
+		await this.retrospectivesRepository.updateOneRetrospective(query, update);
```
Comment (Maghwyn) [inline back-end/src/common/gateways/retrospective.global.gateway.ts:115]: If i emit this event, the retro end early.
Fix it.

File: back-end/src/common/gateways/retrospective.global.gateway.ts [hunk 2]
```diff
@@ -101,4 +101,82 @@ export class RetrospectiveGateway
...
+	}
+	@SubscribeMessage('lock-retro')
+	async lockCurrentRetro(client: AuthSocket, lock: boolean) {
+		client.to(client.roomId).emit('lock-retro', lock);
+		const query = { slug: client.roomId };
+		const update = {$set: { isLocked: lock }}
+
+		await this.retrospectivesRepository.updateOneRetrospective(query, update);
```
Comment (Maghwyn) [inline back-end/src/common/gateways/retrospective.global.gateway.ts]: Same, there's not validation it's open bar.

File: back-end/src/common/gateways/retrospective.global.gateway.ts [hunk 3]
```diff
@@ -101,4 +101,82 @@ export class RetrospectiveGateway
...
+	//@@@@@@@@@@@ TIMER SECTION @@@@@@@@@@@@@@@
+
+	@SubscribeMessage('start-timer')
+	async startTimer(client: AuthSocket) {
+		client.to(client.roomId).emit('start-timer');
+
+		const query = { slug: client.roomId };
+		const update = {$set: { isTimerRunning: true }};
+
+		await this.retrospectivesRepository.updateOneRetrospective(query, update);
```
Comment (Maghwyn) [inline back-end/src/common/gateways/retrospective.global.gateway.ts:139]: So i can re-open the retro basically

File: back-end/src/common/gateways/retrospective.global.gateway.ts [hunk 4]
```diff
@@ -101,4 +101,82 @@ export class RetrospectiveGateway
...
+	async resetTimer(client: AuthSocket) {
+		const query = { _id: client.user.id as ObjectId };
+		const user = await this.usersRepository.findOne(query);
+		const queryRetro = { slug: client.roomId };
+		const currentRetro = await this.retrospectivesRepository.findOne(queryRetro);
+
+		const update = {$set: { endedAt: null, isRetroEnded: false}}
+
+
+		if ((user.role === 2 || user.role === 3) && currentRetro.isRetroEnded) {
```
Comment (Maghwyn) [inline back-end/src/common/gateways/retrospective.global.gateway.ts]: Use the enumeration not number

File: back-end/src/common/gateways/retrospective.global.gateway.ts [hunk 5]
```diff
@@ -101,4 +101,82 @@ export class RetrospectiveGateway
...
+			await this.retrospectivesRepository.updateOneRetrospective(currentRetro, update);
+			client.to(client.roomId).emit('reset-timer');
+		}
+		if ((user.role === 1 || user.role === 2 || user.role === 3) && !currentRetro.isRetroEnded) {
```
Comment (Maghwyn) [inline back-end/src/common/gateways/retrospective.global.gateway.ts]: Same

**Review by Maghwyn — CHANGES_REQUESTED** [review]

File: back-end/src/common/gateways/retrospective.global.gateway.ts (hunk 1 above)
Comment (Kevw94) [reply back-end/src/common/gateways/retrospective.global.gateway.ts:115]: Cet event est reçu que lorsque la rétro arrive à 0 donc, je ne vois pas comment on peut l'emit avant ?

File: back-end/src/common/gateways/retrospective.global.gateway.ts (hunk 3 above)
Comment (Kevw94) [reply back-end/src/common/gateways/retrospective.global.gateway.ts:139]: Yes

File: back-end/src/common/gateways/retrospective.global.gateway.ts (hunk 1 above)
Comment (Maghwyn) [reply back-end/src/common/gateways/retrospective.global.gateway.ts:115]: Not fixed, but merging.

**Pull request merged**